File: Day4.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

std::vector<std::string> get_lines() {
  std::vector<std::string> lines = {};
  std::string line = "";

  std::ifstream input("input.txt");
  while (input >> line) lines.push_back(line);
  input.close();

  return lines;
}

int memory(std::string line) {
  int acc = 0;
  for (char c : line) acc += c - (c >= 'A' && c <= 'Z' ? 64 : 48);

  return acc;
}

std::string lossy_compress(std::string line) {
  int n = line.size() / 10;

  return line.substr(0, n)
    + std::to_string(line.size() - 2 * n)
    + line.substr(line.size() - n);
}

std::string lossless_compress(std::string line) {
  std::string out = "";

  while (!line.empty()) {
    const char c = line[0];
    size_t n = 0;
    while (line[n] == c) n++;

    out += std::to_string(n) + c;
    line.erase(0, n);
  }

  return out;
}

int main() {
  const std::vector<std::string> lines = get_lines();

  int total1 = 0, total2 = 0, total3 = 0;

  for (std::string line : lines) {
    total1 += memory(line);
    total2 += memory(lossy_compress(line));
    total3 += memory(lossless_compress(line));
  }
    
  std::cout << "Part 1 answer: " << total1 << '\n';
  std::cout << "Part 2 answer: " << total2 << '\n';
  std::cout << "Part 3 answer: " << total3;
}

File: Day4/Day4.rs
use std::fs;

fn main() {
  let input = fs::read_to_string("input.txt").unwrap();
  let lines: Vec<_> = input.split_whitespace().collect();

  let total1 = lines.iter().fold(0, |acc, &line| acc + memory(line));
  println!("Part 1 answer: {total1}");

  let total2 = lines.iter().fold(0, |acc, &line| acc + memory(&lossy_compress(line)));
  println!("Part 2 answer: {total2}");

  let total3 = lines.iter().fold(0, |acc, &line| acc + memory(&lossless_compress(line)));
  println!("Part 3 answer: {total3}");
}

fn memory(line: &str) -> u32 {
  line.chars().fold(0, |acc, c| {
    acc + match c.to_digit(10) {
      Some(i) => i,
      None => (c as u8 - 64) as u32
    }
  })
}

fn lossy_compress(line: &str) -> String {
  let n = line.len() / 10;

  format!("{}{}{}",
    &line[..n],
    line.len() - 2 * n,
    &line[line.len()-n..]
  )
}

fn lossless_compress(line: &str) -> String {
  let mut s = String::from(line);
  let mut out = String::new();

  while !s.is_empty() {
    let c = s.chars().next().unwrap();
    let n = s.drain(
        ..s.chars().position(|e| e != c)
        .unwrap_or(s.len())
      ).collect::<String>()
      .len();

    out += format!("{n}{c}").as_str();
  }

  out
}
